thehive/app/org/thp/thehive/services/ShareSrv.scala

```diff
@@ -90,19 +91,44 @@ class ShareSrv @Inject() (
 //    caseSrv.get(`case`).in[ShareCase].filter(_.in[OrganisationShare])._id.getOrFail().flatMap(remove(_)) // FIXME add organisation ?
 
   def remove(shareId: EntityIdOrName)(implicit graph: Graph, authContext: AuthContext): Try[Unit] =
+    unshareCase(shareId).flatMap(_ => Try(get(shareId).remove()))
+
+  def cascadeRemove(shareId: EntityIdOrName)(implicit graph: Graph, authContext: AuthContext): Try[Unit] = {
+    for {
+      organisation <- organisationSrv.getOrFail(authContext.organisation)
+      share        <- getOrFail(shareId)
+
+      tasks          <- Try(get(share).tasks)
+      tasksToUnshare <- Try(tasks.clone().filter(_.inE[ShareTask].count.is(P.gt(1))).toSeq)
+      tasksToDelete  <- Try(tasks.filter(_.inE[ShareTask].count.is(P.eq(1))).toSeq)
+      _              <- tasksToUnshare.toTry(unshareTask(_, organisation))
+      _              <- tasksToDelete.toTry(taskSrv.cascadeRemove(_, share))
+
+      obs          <- Try(get(share).observables)
+      obsToUnshare <- Try(obs.clone().filter(_.inE[ShareObservable].count.is(P.gt(1))).toSeq)
+      obsToDelete  <- Try(obs.filter(_.inE[ShareObservable].count.is(P.eq(1))).toSeq)
+      _            <- obsToUnshare.toTry(unshareObservable(_, organisation))
+      _            <- obsToDelete.toTry(observableSrv.cascadeRemove(_, share))
+
+      cazeDelete   <- Try(get(share).`case`.filter(_.inE[ShareCase].count.is(P.eq(1))).toSeq)
```

Comment:
cazeDelete is empty or contains only one element => use Option instead of Seq for clarity